homeassistant/components/mqtt/client.py

```diff
@@ -372,10 +372,10 @@ def __init__(
         self.config_entry = config_entry
         self.conf = conf
 
-        self._simple_subscriptions: defaultdict[str, set[Subscription]] = defaultdict(
-            set
+        self._simple_subscriptions: defaultdict[str, list[Subscription]] = defaultdict(
+            list
         )
-        self._wildcard_subscriptions: set[Subscription] = set()
+        self._wildcard_subscriptions: list[Subscription] = []
```

Comment:
Why is this changed?

Reply:
May be a list works because we do need to resubscribe when subscribing to the same topic? And a set does not triigger the resubscribe causing the second entity to miss some updates?

Reply:
I made a mistake and sent some changes that didn't belong in the MR. It's now fixed. I only need to revert the change from set to list for the _wildcard_subscriptions and _simple_subscriptions variables.

I do suspect this problem. I have little experience of the HomeAssistant technical stack. I've simply realized that reverting this change solves the problem.

I'll try to create some tests, but if you have some time to help me with this, that would be great.

I don't know if you've had time to see the issue I linked to on Z2M (see PR description), but I'm not the only one with this problem.

Reply:
I have asked @bdraco to have a look as well. I have limited time to have a look at issues.

Reply:
Maybe our `Subscription` object isn't unique enough.  For testing we could raise if it already exists and than rework the `Subscription` object to ensure its unique

Reply:
Assuming its unexpectedly matching an existing subscription (not sure how it could since `HassJob` should always be new), we could set `eq=False` so it compares on `id` instead.

```diff
diff --git a/homeassistant/components/mqtt/client.py b/homeassistant/components/mqtt/client.py
index 2ebd105b432..ac05fb2b487 100644
--- a/homeassistant/components/mqtt/client.py
+++ b/homeassistant/components/mqtt/client.py
@@ -260,7 +260,7 @@ def subscribe(
     return remove
 
 
-@dataclass(slots=True, frozen=True)
+@dataclass(slots=True, frozen=True, eq=False)
 class Subscription:
     """Class to hold data about an active subscription."""
 
```

Reply:
We really need to understand what is going on here before moving a fix forward as its going to get broken again in the future if we don't understand the problem.  We need a test case for this issue.

Reply:
Hello @bdraco `eq=False` on the latest version of the dev branch and that doesn't solve the problem either. I'll try to provide you with logs and tests. It would also have been a good thing if tests had been added on #118088